### PR TITLE
Fix obsolete X509Certificate2 Errors

### PR DESCRIPTION
**Pull request by ellahathaway**
Fixes https://github.com/dotnet/source-build/issues/4494

[Test build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2491390&view=results)

My local build worked with thes changes, but I'm verifying with the source-build-lite pipeline as well.

### CODE REVIEW

**Comment by ellahathaway** [comment]
It looks like we may have to pull in the latest from runtime in order for this to work at the repo level.

**Comment by vcsjones** [comment]
👋 This is an interesting case as, currently, there is no replacement API for what you are using `X509Certificate2` for. You are likely going to need to suppress the SYSLIB0057 obsoletion as `X509CertificateLoader` does not handle authenticode inputs.

cc @bartonjs  - this is another example of needing to suppress the obsoletion.

**Comment by akoeplinger** [comment]
See https://github.com/dotnet/docs/issues/41662#issuecomment-2214730902 for an example.

**Comment by ellahathaway** [comment]
@vcsjones @akoeplinger - Thanks for your help! I appreciate it. I've updated the PR given the guidance

**Review by ViktorHofer — APPROVED** [review]

**Review by chcosta — DISMISSED** [review]

**Comment by ellahathaway** [comment]
Stage 2 build passed for arcade with these changes. I will have to investigate the arcade-pr failures after I finish with release work.

**Comment by bartonjs** [comment]
> We'd need to wrap it in #if NET9_0_OR_GREATER ...

Or add a package reference to Microsoft.Bcl.Cryptography (the back-compat package)

**Review by bartonjs — APPROVED** [review]

**Comment by ellahathaway** [comment]
One of the tests was failing because the filePath did not contain a certificate. This resulted in GetCertContentType() throwing an exception. As a fix, I placed the check for the cert type in the try-catch block.

File: src/Microsoft.DotNet.SignTool/src/VerifySignatures.cs [hunk 1]
```diff
@@ -99,20 +99,18 @@ internal static bool IsSignedContainer(string fullPath, string tempDir, string t
...
             {
-                // X509CertificateLoader does not handle authenticode inputs
-                // so we have to suppress the X509Certificate2 ctor obsoletion
-                #pragma warning disable SYSLIB0057
+                // We later suppress SYSLIB0057 because X509CertificateLoader does not handle authenticode inputs
+                // so we should verify that the certificate is authenticode before using X509Certificate2.CreateFromSignedFile
+                var certContentType = X509Certificate2.GetCertContentType(fullPath);
+                if (certContentType != X509ContentType.Authenticode)
+                {
+                    throw new CryptographicException($"Unexpected certificate content type, got '{certContentType}' instead of Authenticode.");
```
Comment (bartonjs) [inline src/Microsoft.DotNet.SignTool/src/VerifySignatures.cs:110]: While throwing would be the normal approach, you're in a try/catch that will catch this exception, eat it (no logging or what have you), and return false.

You could just `return false;` here.

File: src/Microsoft.DotNet.SignTool/src/VerifySignatures.cs [hunk 2]
```diff
@@ -101,8 +102,18 @@ internal static bool IsDigitallySigned(string fullPath)
             X509Certificate2 certificate;
             try
             {
+                // We later suppress SYSLIB0057 because X509CertificateLoader does not handle authenticode inputs
+                // so we should verify that the certificate is authenticode before using X509Certificate2.CreateFromSignedFile
+                var certContentType = X509Certificate2.GetCertContentType(fullPath);
+                if (certContentType != X509ContentType.Authenticode)
+                {
+                    throw new CryptographicException($"Unexpected certificate content type, got '{certContentType}' instead of Authenticode.");
```
Comment (akoeplinger) [inline src/Microsoft.DotNet.SignTool/src/VerifySignatures.cs:110]:
```suggestion
                    return false;
```

**Review by akoeplinger — APPROVED** [review]

File: src/Microsoft.DotNet.SignTool/src/VerifySignatures.cs (hunk 1 above)
Comment (akoeplinger) [reply src/Microsoft.DotNet.SignTool/src/VerifySignatures.cs]: Applied that change, thanks.

**Review by ViktorHofer — APPROVED** [review]

**Pull request merged**